src/dataTransfer/dataTransferClip.ts

```diff
@@ -5,41 +5,61 @@ import DataTransferFrame from './dataTransferFrame'
 
 export default class DataTransferClip extends DataTransfer {
 	public readonly clipIndex: number // 0 or 1
-	public readonly frames: Array<DataTransferFrame>
+	public readonly frames: Generator<DataTransferFrame> | AsyncGenerator<DataTransferFrame>
 	public readonly name: string
-	public curFrame = 0
+	public curFrame: DataTransferFrame | undefined
+	private numFrames = 0
+	private started = false
 
-	constructor(clipIndex: number, name: string, frames: Array<DataTransferFrame>) {
+	constructor(
+		clipIndex: number,
+		name: string,
+		frames: Generator<DataTransferFrame> | AsyncGenerator<DataTransferFrame>
+	) {
 		super(0, 1 + clipIndex)
 
 		this.clipIndex = clipIndex
 		this.name = name
 		this.frames = frames
 	}
 
-	public start(): Commands.ISerializableCommand[] {
+	public async start(): Promise<Commands.ISerializableCommand[]> {
 		const commands: Commands.ISerializableCommand[] = []
 		commands.push(new Commands.MediaPoolClearClipCommand(this.clipIndex))
-		this.frames[this.curFrame].state = Enums.TransferState.Locked
-		commands.push(...this.frames[this.curFrame].start())
+		this.curFrame = await (await this.frames.next()).value
+		if (!this.curFrame) {
+			throw new Error('No frames available for transfer')
+		}
+		this.numFrames++
+		this.curFrame.state = Enums.TransferState.Locked
+		commands.push(...this.curFrame.start())
+		this.started = true
 		return commands
 	}
 
-	public handleCommand(command: Commands.IDeserializedCommand): Commands.ISerializableCommand[] {
+	public async handleCommand(command: Commands.IDeserializedCommand): Promise<Commands.ISerializableCommand[]> {
+		if (!this.started) {
+			await this.waitForStart()
```

Comment:
Is this creating a race condition? Does this guarantee that commands will be handled in the order they were received? If not, how can that guarantee be made?

Reply:
Iirc the protocol uses an internal counter to ensure packets are received in order (by packetId), a race condition however may still occur because the commands are emitted as events so while 1 command is being handled, another may bubble up and due to async code create a race condition. It seems a little far fetched but to be certain we could use a promise queue to ensure handling commands in order?

Reply:
It might be, but im not sure its necessary to wait for the start here.
Until the commands returned from start() are sent then anything fed into handleCommand cant be for this transfer, and this.started will be made true before any of those commands are sent. So I think this should either be a `return []` or maybe error in some way but im not sure if that should error back to the atem or log something

Reply:
In my previous testing I found that it was necessary to wait for start, because some lock-related commands were coming in to `handleCommand` before `start` had finished executing. However, since adding a `p-queue` to the `DataTransferManager`, I've been unable to replicate that issue and have removed this code for now.